### PR TITLE
CAT-1869 DragNDrop for Sprites, Looks and Sounds

### PR DESCRIPTION
**Pull request by loibi93**
Context Menus are removed (also Tests are fixed to only use ActionMode when needed). DragNDrop is not implemented to be used in Backpack (since it is not needed there, also confirmed by UX Team).

### CODE REVIEW

**Comment by loibi93** [comment]
ok to test

File: catroid/src/org/catrobat/catroid/ui/DynamicListView.java [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+
+public class DynamicListView extends ListView {
+
+	private static final int SMOOTH_SCROLL_AMOUNT_AT_EDGE = 15;
+	private static final int MOVE_DURATION = 100;
+	private static final int LINE_THICKNESS = 15;
+
+	public List dataList;
+	private boolean forSpriteList = false;
+
```
Comment (jonny911) [inline catroid/src/org/catrobat/catroid/ui/DynamicListView.java]: Is there a reason for the blank lines?

File: catroidTest/src/org/catrobat/catroid/uitest/ui/fragment/LookFragmentTest.java [hunk 2]
```diff
@@ -819,84 +812,52 @@ public void testRenameLookContextMenu() {
...
-		solo.sleep(TIME_TO_WAIT);
 		solo.goBack();
 		solo.clickOnText(solo.getString(R.string.backgrounds));
 
-		assertEquals("Look didn't move top (testMoveLookToTop 1)", SECOND_TEST_LOOK_NAME, getLookName(0));
-		assertEquals("Look didn't move top (testMoveLookToTop 2)", FIRST_TEST_LOOK_NAME, getLookName(1));
-	}
+		assertEquals("Wrong List before DragAndDropTest", lookDataList.get(0).getLookName(), FIRST_TEST_LOOK_NAME);
+		assertEquals("Wrong List before DragAndDropTest", lookDataList.get(1).getLookName(), SECOND_TEST_LOOK_NAME);
+		assertEquals("Wrong List before DragAndDropTest", lookDataList.get(2).getLookName(), "TestLook0");
```
Comment (jonny911) [inline catroidTest/src/org/catrobat/catroid/uitest/ui/fragment/LookFragmentTest.java]: Is there a reason, why TestLookx is hardcoded?

File: catroidTest/src/org/catrobat/catroid/uitest/ui/fragment/NfcTagFragmentTest.java [hunk 3]
```diff
@@ -832,108 +833,52 @@ public void testSelectAllActionModeButton() {
...
-		solo.goBack();
-		solo.clickOnText(solo.getString(R.string.backgrounds));
-
-		assertEquals("Tag didn't move top (testMoveTagToTop 1)", SECOND_TEST_TAG_NAME, getTagName(0));
-		assertEquals("Tag didn't move top (testMoveTagToTop 2)", FIRST_TEST_TAG_NAME, getTagName(1));
-		assertEquals("Tag didn't move top (testMoveTagToTop 3)", THIRD_TEST_TAG_NAME, getTagName(2));
-	}
+	public void testDragAndDropUp() {
+		for (int i = 0; i < 2; i++) {
+			addNFCTagWithName("TestNFC" + i);
```
Comment (jonny911) [inline catroidTest/src/org/catrobat/catroid/uitest/ui/fragment/NfcTagFragmentTest.java]: Is there a reason, why TestNFCx is hardcoded?

File: catroidTest/src/org/catrobat/catroid/uitest/util/UiTestUtils.java [hunk 4]
```diff
@@ -1632,7 +1632,7 @@ public static void acceptAndCloseActionMode(Solo solo) {
 		View doneButton = solo.getView(Resources.getSystem().getIdentifier("action_mode_close_button", "id", "android"));
 
 		solo.clickOnView(doneButton);
-		solo.sleep(200);
+		solo.sleep(400);
```
Comment (jonny911) [inline catroidTest/src/org/catrobat/catroid/uitest/util/UiTestUtils.java]: Can this be replaced by solo.waitForxxx()?

File: catroidTest/src/org/catrobat/catroid/uitest/ui/fragment/LookFragmentTest.java (hunk 2 above)
Comment (loibi93) [inline catroidTest/src/org/catrobat/catroid/uitest/ui/fragment/LookFragmentTest.java]: Yes, as you can see, when the test starts, i add some Looks in a for loop. The names of these looks are composed with TestLook plus a unique number (1, 2 and 3 in this case). So the names are composed at runtime. Same goes for the same issue in NFCTagFragmentTest.

File: catroidTest/src/org/catrobat/catroid/uitest/ui/fragment/NfcTagFragmentTest.java (hunk 3 above)
Comment (loibi93) [inline catroidTest/src/org/catrobat/catroid/uitest/ui/fragment/NfcTagFragmentTest.java]: same as above.

File: catroid/src/org/catrobat/catroid/ui/DynamicListView.java [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+					cellIsMobile = true;
+
+					updateNeighborViewsForID(mobileItemId);
+
+					return true;
+				}
+			};
+
+	private BitmapDrawable getAndAddHoverView(View v) {
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/ui/DynamicListView.java]: Remove newline

File: catroid/src/org/catrobat/catroid/ui/DynamicListView.java [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+	}
+
+	private AdapterView.OnItemLongClickListener onItemLongClickListener =
+			new AdapterView.OnItemLongClickListener() {
+				public boolean onItemLongClick(AdapterView<?> arg0, View arg1, int pos, long id) {
+					if ((pos == 0 && forSpriteList)) {
+						return true;
+					}
+					dismissLongPress = true;
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/ui/DynamicListView.java]: Remove newline

File: catroid/src/org/catrobat/catroid/ui/DynamicListView.java [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+		boolean isBelow = (belowView != null) && (deltaYTotal > belowView.getTop());
+		boolean isAbove = (aboveView != null) && (deltaYTotal < aboveView.getTop());
+
+		if (aboveView != null && (getPositionForView(aboveView) == 0 && isAbove && forSpriteList)) {
+			return;
+		}
+
+		if (isBelow || isAbove) {
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/ui/DynamicListView.java]: Remove newline

File: catroid/src/org/catrobat/catroid/ui/DynamicListView.java [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+	public void setDataList(List data) {
+		dataList = data;
+	}
+
+	public void isForSpriteList() {
+		forSpriteList = true;
+	}
+
+	private AbsListView.OnScrollListener scrollListener = new AbsListView.OnScrollListener() {
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/ui/DynamicListView.java]: Remove newline

File: catroid/src/org/catrobat/catroid/ui/DynamicListView.java [hunk 9]
```diff
@@ -0,0 +1,519 @@
...
+		int range = computeVerticalScrollRange();
+		int hoverViewTop = r.top;
+		int hoverHeight = r.height();
+
+		if (hoverViewTop <= 0 && offset > 0) {
+			smoothScrollBy(-smoothScrollAmountAtEdge, 0);
+			return true;
+		}
+
+		if (hoverViewTop + hoverHeight >= height && (offset + extent) < range) {
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/ui/DynamicListView.java]: Use an boolean value with an meaningful name for better readability

File: catroid/src/org/catrobat/catroid/ui/adapter/LookAdapter.java [hunk 10]
```diff
@@ -33,16 +33,58 @@
...
 import java.util.List;
 
 public class LookAdapter extends LookBaseAdapter implements ActionModeActivityAdapterInterface,
 		LookController.OnBackpackLookCompleteListener {
 
+	private static final int INVALID_ID = -1;
+
 	private LookFragment lookFragment;
 
+	HashMap<LookData, Integer> idMap = new HashMap<LookData, Integer>();
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/ui/adapter/LookAdapter.java]: insert private

File: catroid/src/org/catrobat/catroid/ui/adapter/NfcTagAdapter.java [hunk 11]
```diff
@@ -32,16 +32,49 @@
...
 import java.util.Iterator;
 import java.util.List;
 
 public class NfcTagAdapter extends NfcTagBaseAdapter implements ActionModeActivityAdapterInterface {
 
+	private static final int INVALID_ID = -1;
+
 	private NfcTagFragment nfcTagFragment;
 
+	HashMap<NfcTagData, Integer> idMap = new HashMap<NfcTagData, Integer>();
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/ui/adapter/NfcTagAdapter.java]: insert private

File: catroid/src/org/catrobat/catroid/ui/adapter/SoundAdapter.java [hunk 12]
```diff
@@ -35,16 +35,49 @@
...
 import java.util.Iterator;
 import java.util.List;
 
 public class SoundAdapter extends SoundBaseAdapter implements ActionModeActivityAdapterInterface, SoundController.OnBackpackSoundCompleteListener {
 
+	private static final int INVALID_ID = -1;
+
 	private SoundFragment soundFragment;
 
+	HashMap<SoundInfo, Integer> idMap = new HashMap<SoundInfo, Integer>();
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/ui/adapter/SoundAdapter.java]: insert private

File: catroid/src/org/catrobat/catroid/ui/adapter/SpriteAdapter.java [hunk 13]
```diff
@@ -41,17 +43,51 @@
...
+import java.util.HashMap;
 import java.util.List;
 
 public class SpriteAdapter extends SpriteBaseAdapter implements ActionModeActivityAdapterInterface {
 
+	private static final int INVALID_ID = -1;
+
 	SpritesListFragment spritesListFragment;
 
+	HashMap<Sprite, Integer> idMap = new HashMap<Sprite, Integer>();
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/ui/adapter/SpriteAdapter.java]: insert private

File: catroid/src/org/catrobat/catroid/ui/controller/SoundController.java [hunk 14]
```diff
@@ -253,6 +254,25 @@ public void onClick(View view) {
 			}
 		}
 		holder.soundFragmentButtonLayout.setOnClickListener(listItemOnClickListener);
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/ui/controller/SoundController.java]: remove newline

**Comment by loibi93** [comment]
@thomasmauerhofer fixed issues ;)

**Comment by loibi93** [comment]
Can this be merged now ? @thmq

**Comment by loibi93** [comment]
ok to test

**Comment by loibi93** [comment]
Rebased and squashed to one commit, approved by UX and code reviewed, should be good to go !

CatroidCheck: https://jenkins.catrob.at/view/LeeroyCatroid/job/CatroidCheck/217/
CatroidSinglePackageEmulatorTest: https://jenkins.catrob.at/view/LeeroyCatroid/job/CatroidSinglePackageEmulatorTest/1803/

**Comment by p4p4** [comment]
@loibi93 how to delete a sprite? is there a new way to do this? thanks! :)

**Pull request merged**